### ISSUE TITLE
[no-unused-vars] False positive with some arrow functions

### ISSUE THREAD

**Issue by anna-osipova** [typescript-eslint/typescript-eslint#2874]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "plugins": ["@typescript-eslint", "jest", "prettier", "react", "react-hooks", "lodash"],
  "rules": {
   "@typescript-eslint/no-unused-vars": ["error", { "ignoreRestSiblings": true }]
  }
}
```

```TS
export const activeCompanyId = createSelector(
  libraryId,
  libraries,
  (libraryId, libs) => libs.find((lib) => lib.id === libraryId)?.company_id
);
```

**Expected Result**

No errors - all variables clearly being used

**Actual Result**

error  'libraryId' is defined but never used  @typescript-eslint/no-unused-vars
error  'libs' is defined but never used       @typescript-eslint/no-unused-vars

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.1` |
| `@typescript-eslint/parser`        | `4.9.1` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `7.15.0` |
| `node`                             | `12.20.0` |

**Comment by tryggvigy** [comment]
@anna-osipova this bug has something to do with optional fields. I saw the same issue in createSelector after updating and managed to work around it by making sure the selectors did not take optional arguments, which was ok in my case.

```ts
export const activeCompanyId = createSelector(
  libraryId,
  libraries,
  (libraryId, libs) => libs.find((lib) => lib.id === libraryId)?.company_id
);
```

If that's acceptable for you, try changing the arguments of the `libraryId` and `libraries` selectors from 

```ts
const libraryId = (arg?: whatever) => {...}
```
to
```ts
const libraryId = (arg: whatever | undefined) => {...}
```

Looks like this is still definitely a bug in `typescript-eslint`

**Comment by anna-osipova** [comment]
@tryggvigy You're right, it's something to do with optional values. The problem is `.find` returns a value or null. 

If I change it like below, it works, but it's a lot more lines of code 😔
```TS
export const activeCompanyId = createSelector(libraryId, libraries, (libraryId, libs) => {
  const library = libs.find((lib) => lib.id === libraryId);
  if (library) {
    return library.company_id;
  }
  return null;
});
```

**Comment by bradzacher** [comment]
I'm unable to repro this against master.

![image](https://user-images.githubusercontent.com/7462525/102108627-0295a280-3de8-11eb-894e-ec612a349e68.png)

That generally implies that your versions installed are different to the ones you think are installed.
I recommend that you verify your installation, and worst case delete your `node_modules` folder and reinstall.
